lib/collection/src/collection_manager/holders/segment_holder.rs

```diff
@@ -547,25 +582,32 @@ impl<'s> SegmentHolder {
             &mut RwLockWriteGuard<dyn SegmentEntry>,
             &T,
         ) -> OperationResult<bool>,
+        OD: FnMut(PointIdType, &mut RwLockWriteGuard<dyn SegmentEntry>) -> OperationResult<bool>,
     {
         let _update_guard = self.update_tracker.update();
 
-        // Select what points to update in what segments
-        let segment_points = if all_point_versions {
-            self.all_segments_points(ids)
-        } else {
-            self.latest_segments_points(ids)
-        };
+        let (to_update, to_delete) = self.find_points_to_update_and_delete(ids);
+
+        // Delete old points first
+        for (segment_id, points) in to_delete {
+            let segment = self.get(segment_id).unwrap();
+            let segment_arc = segment.get();
+            let mut write_segment = segment_arc.write();
+
+            for point_id in points {
+                point_delete_operation(point_id, &mut write_segment)?;
+            }
+        }
```

Comment:
Here we now delete older versions before applying operations to the newest point version.